src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
+    Tuple,
+    Unit,
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct StructDef {
     /// Fields, not including ctor
     pub fields: Vec<StructField>,
     /// ID of the constructor. This is only used for tuple- or enum-like
     /// structs.
-    pub ctor_id: Option<NodeId>,
+    pub id: NodeId,
```

Comment:
`Variant in Enum |...|Used |Used` will require some additional refactoring, because now both `Used` successfully share the same id. I'll look at it and at `with_fields(|field| { /* ... */ })` tomorrow.
(I've also updated the table with `NodeId in Item` column.)